### ISSUE TITLE
Host interface forward compatibility

### ISSUE THREAD

**Issue by lann** [fermyon/spin#1124]
For language (toolchains) without reliable dead code elimination, Spin SDKs are faced with a forward-compatibility problem when adding new interface imports: if a host is missing any imported functions - even if they aren't actually used by the guest - linking will fail.

There are two related situations where this comes up:

- Adding a new interface: `outbound-mysql`
- Adding a method to an existing interface: `outbound-redis::del` (the actual motivating example here)

Several approaches to address this:

1. Set a policy that all features must be released in Spin before being released for SDKs. To minimize confusion we probably wouldn't even want to _merge_ features to SDKs before Spin.

1. Split major features into separate SDK packages and have users opt-in to each feature. Each package would have a minimum required Spin version. This doesn't solve forward-compatibility for additions _within_ an interface (e.g. `outbound-redis::del`), but could be combined with a function stubbing approach like below.

1. Create a feature detection system for Spin imports. For example, each Spin feature interface could have a "dummy" `spin-feature-detection: func()`. For any import module with this function, Spin would set a trapping stub function for any unknown import functions. This is roughly what [`Linker::define_unknown_imports_as_traps`](https://docs.rs/wasmtime/5.0.0/wasmtime/struct.Linker.html#method.define_unknown_imports_as_traps) does, but would be scoped to these individual import modules.

**Comment by dicej** [comment]
Philosophically, we'll need to decide, as we "fully embrace" the component model, what a guest component import really means.  If it means "I really need this, and don't bother trying to instantiate me if you can't provide it", then trapping stubs are never appropriate, and we have to address this from a different angle.  If it means "I might need this -- we'll know for sure at runtime", then trapping stubs are absolutely a fine approach.

I do like the `spin-feature-detection` approach, because the guest can use it to clearly communicate "all imports from this module are optional", i.e. it's explicitly opting in to the second interpretation above.  Guests which don't use `spin-feature-detection` can therefore be safely assumed to be opting into the first interpretation.

**Comment by lann** [comment]
Yeah, I'm hoping we can have something nicer with the component model, but I'm not sure that optional imports et al are even slated for initial release so I think it would be good to have an interim solution, even if ends up only applying to modules.

**Comment by itowlson** [comment]
@dicej The absolutist perspective seems to push us to option 2 (or to something even more granular) - if an import means "I _must_ have this", then each language package should declare only the absolute minimum set of imports (unless we assume _iron clad guaranteed_ unused import elimination), and we'd have a package per interface or so.  As Lann notes, though, that leaves us with a problem for adding stuff to an interface - it would be absurd to have a language package for each individual Redis function, but if a component doesn't use `del` then it shouldn't import it.

In practice, multiple granular packages likely leaves us with a problem long before then, because users will be inclined to "eh just bang in all the Spin libraries, it works and saves faffing around."  (I can well imagine pressure for the templates to import everything on the basis of discoverability.)  It's a convenience issue rather than a correctness issues, but we know we developers do love us some convenience...

**Comment by lann** [comment]
I'm not sure how useful it would be, but a slight variation of the above proposal could allow guests to dynamically detect support for specific functions, e.g.

```yaml
spin-feature-detection: func(func-name: string) -> bool
```
